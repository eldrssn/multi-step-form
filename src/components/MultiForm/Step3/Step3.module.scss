@import '/src/styles/breakpoints.scss';
@import '/src/styles/colors.scss';

.buttons {
	position: absolute;
	bottom: 0;

	width: 100%;
	display: flex;
	justify-content: space-between;
}

.addons {
	display: flex;
	flex-direction: column;
	gap: 16px;
}

.info {
  flex-grow: 1;
}

@media screen and (max-width: $mobile) {

  .addons {
    gap: 10px;
  }

  .buttons {
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: $white;
    padding: 15px;
  }
}