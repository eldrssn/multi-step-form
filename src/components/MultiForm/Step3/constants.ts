export const addOns: Record<string, string>[] = [
  {
    title: 'Online service',
    name: 'service',
    description: 'Access to multiplayer games',
    priceYearly: '10',
    priceMonthly: '1',
  },
  {
    title: 'Larger storage',
    name: 'storage',
    description: 'Extra 1TB of cloud save',
    priceYearly: '20',
    priceMonthly: '2',
  },
  {
    title: 'Customizable Profile',
    name: 'profile',
    description: 'Custom theme on your profile',
    priceYearly: '20',
    priceMonthly: '2',
  },
];
