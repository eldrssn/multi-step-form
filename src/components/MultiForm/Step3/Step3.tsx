import { FC, useState } from 'react';

import { FormHeader } from '@components/ui/FormHeader';
import { Button } from '@components/ui/Button';
import { AddOn } from '@components/ui/AddOn';

import { TStep, TStep3 } from '../types';
import { addOns } from './constants';
import styles from './Step3.module.scss';

const Step3: FC<TStep> = ({ handleNext, handleBack, data }) => {
  const [selectedAddOns, setSelectedAddOns] = useState<TStep3>(
    data?.step3Data || {}
  );

  const isYearly = !!data.step2Data?.isYearly;

  const handleCheckboxChange = (addOnTitle: string) => {
    setSelectedAddOns({
      ...selectedAddOns,
      [addOnTitle]: !selectedAddOns[addOnTitle],
    });
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    handleNext && handleNext({ step3Data: { ...selectedAddOns } });
  };
  return (
    <div>
      <FormHeader
        title="Pick add-ons"
        description="Add-ons help enhance your gaming experience."
      />
      <form onSubmit={handleSubmit}>
        <div className={styles.addons}>
          {addOns.map((addOn) => (
            <AddOn
              isYearly={isYearly}
              key={addOn.name}
              addOn={addOn}
              onChange={() => handleCheckboxChange(addOn.name)}
              isChecked={selectedAddOns[addOn.name] || false}
            />
          ))}
        </div>
        <div className={styles.buttons}>
          <Button action="back" onClick={handleBack} />
          <Button action="next" type="submit" />
        </div>
      </form>
    </div>
  );
};

export { Step3 };
