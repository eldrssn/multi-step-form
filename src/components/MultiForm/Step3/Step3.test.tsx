import { fireEvent, render } from '@testing-library/react';
import { Step3 } from './Step3';

const renderStep3 = (props = {}) => {
  const defaultProps = {
    handleNext: jest.fn(),
    handleBack: jest.fn(),
    data: { step1Data: null, step2Data: null, step3Data: null },
    ...props,
  };

  return render(<Step3 {...defaultProps} />);
};

describe('Step3', () => {
  test('renders Step3 component correctly', () => {
    const { getByText, getAllByTestId, container } = renderStep3();

    expect(getByText('Pick add-ons')).toBeInTheDocument();
    expect(
      getByText('Add-ons help enhance your gaming experience.')
    ).toBeInTheDocument();
    expect(getAllByTestId('addon')).toHaveLength(3);
    expect(container).toMatchSnapshot();
  });

  test('selecting an add-on updates state', () => {
    const { getAllByTestId } = renderStep3();

    const addonCheckbox = getAllByTestId('addon')[1];
    fireEvent.click(addonCheckbox);

    expect(addonCheckbox).toHaveClass('checked');
  });

  test('submits the form with selected add-ons', async () => {
    const handleNextMock = jest.fn();
    const { getAllByTestId, getByText } = renderStep3({
      handleNext: handleNextMock,
    });

    const addonCheckbox = getAllByTestId('addon')[0];
    fireEvent.click(addonCheckbox);

    fireEvent.click(getByText(/next step/i));

    expect(handleNextMock).toHaveBeenCalledWith({
      step3Data: {
        service: true,
      },
    });
  });

  test('clicking "Back" button calls handleBack', () => {
    const handleBackMock = jest.fn();
    const { getByText } = renderStep3({ handleBack: handleBackMock });

    fireEvent.click(getByText(/Go Back/i));

    expect(handleBackMock).toHaveBeenCalled();
  });
});
