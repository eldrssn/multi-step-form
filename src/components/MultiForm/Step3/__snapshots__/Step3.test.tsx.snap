// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step3 renders Step3 component correctly 1`] = `
<div>
  <div>
    <h1
      class="title"
    >
      Pick add-ons
    </h1>
    <p
      class="description"
    >
      Add-ons help enhance your gaming experience.
    </p>
    <form>
      <div
        class="addons"
      >
        <label
          class="addon"
          data-testid="addon"
        >
          <input
            type="checkbox"
          />
          <span
            class="checkmark"
          />
          <div
            class="info"
          >
            <p
              class="title"
            >
              Online service
            </p>
            <p
              class="description"
            >
              Access to multiplayer games
            </p>
          </div>
          <p
            class="price"
          >
            +$
            1
            /
            mo
          </p>
        </label>
        <label
          class="addon"
          data-testid="addon"
        >
          <input
            type="checkbox"
          />
          <span
            class="checkmark"
          />
          <div
            class="info"
          >
            <p
              class="title"
            >
              Larger storage
            </p>
            <p
              class="description"
            >
              Extra 1TB of cloud save
            </p>
          </div>
          <p
            class="price"
          >
            +$
            2
            /
            mo
          </p>
        </label>
        <label
          class="addon"
          data-testid="addon"
        >
          <input
            type="checkbox"
          />
          <span
            class="checkmark"
          />
          <div
            class="info"
          >
            <p
              class="title"
            >
              Customizable Profile
            </p>
            <p
              class="description"
            >
              Custom theme on your profile
            </p>
          </div>
          <p
            class="price"
          >
            +$
            2
            /
            mo
          </p>
        </label>
      </div>
      <div
        class="buttons"
      >
        <button
          class="button button_back"
          type="button"
        >
          Go Back
        </button>
        <button
          class="button button_next"
          type="submit"
        >
          Next Step
        </button>
      </div>
    </form>
  </div>
</div>
`;
